File: python/unified_channel/adapters/mattermost.py
"""Mattermost adapter — WebSocket + REST API.

Requires: pip install websockets httpx
"""

from __future__ import annotations

import asyncio
import json
import logging
from datetime import datetime
from typing import AsyncIterator

import httpx
import websockets

from ..adapter import ChannelAdapter
from ..types import (
    ChannelStatus,
    ContentType,
    Identity,
    MessageContent,
    OutboundMessage,
    UnifiedMessage,
)

logger = logging.getLogger(__name__)


class MattermostAdapter(ChannelAdapter):
    """Mattermost adapter using WebSocket for events and REST for sending."""

    channel_id = "mattermost"

    def __init__(
        self,
        url: str,
        token: str,
        *,
        allowed_channel_ids: set[str] | None = None,
        command_prefix: str = "/",
    ) -> None:
        self._url = url.rstrip("/")
        self._token = token
        self._allowed_channels = allowed_channel_ids
        self._prefix = command_prefix
        self._queue: asyncio.Queue[UnifiedMessage] = asyncio.Queue()
        self._connected = False
        self._last_activity: datetime | None = None
        self._ws: websockets.WebSocketClientProtocol | None = None
        self._http: httpx.AsyncClient | None = None
        self._bot_user_id: str | None = None
        self._ws_task: asyncio.Task | None = None

    async def connect(self) -> None:
        self._http = httpx.AsyncClient(
            base_url=f"{self._url}/api/v4",
            headers={"Authorization": f"Bearer {self._token}"},
            timeout=30,
        )
        me = await self._http.get("/users/me")
        me.raise_for_status()
        self._bot_user_id = me.json()["id"]

        ws_url = self._url.replace("http", "ws") + "/api/v4/websocket"
        self._ws = await websockets.connect(ws_url)
        await self._ws.send(json.dumps({
            "seq": 1, "action": "authentication_challenge",
            "data": {"token": self._token},
        }))

        self._connected = True
        self._ws_task = asyncio.create_task(self._listen())
        logger.info("mattermost connected: %s", self._bot_user_id)

    async def disconnect(self) -> None:
        self._connected = False
        if self._ws_task:
            self._ws_task.cancel()
        if self._ws:
            await self._ws.close()
        if self._http:
            await self._http.aclose()
        logger.info("mattermost disconnected")

    async def receive(self) -> AsyncIterator[UnifiedMessage]:
        while self._connected:
            try:
                msg = await asyncio.wait_for(self._queue.get(), timeout=1.0)
                yield msg
            except asyncio.TimeoutError:
                continue

    async def send(self, msg: OutboundMessage) -> str | None:
        if not self._http:
            raise RuntimeError("mattermost not connected")
        payload: dict = {"channel_id": msg.chat_id, "message": msg.text}
        if msg.reply_to_id:
            payload["root_id"] = msg.reply_to_id
        resp = await self._http.post("/posts", json=payload)
        self._last_activity = datetime.now()
        if resp.status_code == 201:
            return resp.json().get("id")
        logger.error("mattermost send failed: %d", resp.status_code)
        return None

    async def get_status(self) -> ChannelStatus:
        return ChannelStatus(
            connected=self._connected, channel="mattermost",
            account_id=self._bot_user_id, last_activity=self._last_activity,
        )

    async def _listen(self) -> None:
        try:
            async for raw in self._ws:
                try:
                    event = json.loads(raw)
                except json.JSONDecodeError:
                    continue
                if event.get("event") == "posted":
                    await self._process_post(event)
        except (websockets.ConnectionClosed, asyncio.CancelledError):
            self._connected = False

    async def _process_post(self, event: dict) -> None:
        data = json.loads(event.get("data", {}).get("post", "{}"))
        if data.get("user_id") == self._bot_user_id:
            return
        channel_id = data.get("channel_id", "")
        if self._allowed_channels and channel_id not in self._allowed_channels:
            return

        text = data.get("message", "")
        self._last_activity = datetime.now()

        if text.startswith(self._prefix):
            parts = text[len(self._prefix):].split()
            mc = MessageContent(type=ContentType.COMMAND, text=text,
                                command=parts[0] if parts else "", args=parts[1:])
        else:
            mc = MessageContent(type=ContentType.TEXT, text=text)

        msg = UnifiedMessage(
            id=data.get("id", ""), channel="mattermost",
            sender=Identity(id=data.get("user_id", "")),
            content=mc, chat_id=channel_id,
            thread_id=data.get("root_id"), reply_to_id=data.get("root_id"),
            raw=data,
        )
        await self._queue.put(msg)

File: rust/src/adapters/mattermost.rs
//! Mattermost adapter via WebSocket + REST API.

use async_trait::async_trait;
use chrono::Utc;
use reqwest::Client as HttpClient;
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use tokio::sync::Mutex;

use crate::adapter::{ChannelAdapter, MessageHandler};
use crate::error::{Error, Result};
use crate::types::*;

/// Mattermost configuration.
pub struct MattermostConfig {
    /// Server URL (e.g., "https://mattermost.example.com").
    pub server_url: String,
    /// Personal access token or bot token.
    pub access_token: String,
    /// Team ID to listen to (optional; listens to all if not set).
    pub team_id: Option<String>,
}

impl MattermostConfig {
    /// Create with minimal required fields.
    pub fn new(server_url: impl Into<String>, access_token: impl Into<String>) -> Self {
        Self {
            server_url: server_url.into(),
            access_token: access_token.into(),
            team_id: None,
        }
    }
}

#[derive(Deserialize)]
struct MmUser {
    id: String,
    username: String,
    #[serde(default)]
    nickname: String,
}

#[derive(Deserialize)]
struct MmPost {
    id: String,
    channel_id: String,
    user_id: String,
    message: String,
    #[serde(default)]
    root_id: String,
    create_at: i64,
}

#[derive(Serialize)]
struct MmCreatePost {
    channel_id: String,
    message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    root_id: Option<String>,
}

#[derive(Deserialize)]
struct MmCreatePostResponse {
    id: String,
}

/// Mattermost WebSocket event payload.
#[derive(Deserialize)]
struct MmWsEvent {
    event: String,
    data: serde_json::Value,
}

/// Mattermost adapter.
pub struct MattermostAdapter {
    config: MattermostConfig,
    connected: bool,
    last_activity: Option<chrono::DateTime<Utc>>,
    handler: Arc<Mutex<Option<MessageHandler>>>,
    http: HttpClient,
    user_id: Option<String>,
    ws_shutdown: Option<tokio::sync::oneshot::Sender<()>>,
}

impl MattermostAdapter {
    /// Create a new Mattermost adapter.
    pub fn new(config: MattermostConfig) -> Self {
        Self {
            config,
            connected: false,
            last_activity: None,
            handler: Arc::new(Mutex::new(None)),
            http: HttpClient::new(),
            user_id: None,
            ws_shutdown: None,
        }
    }

    /// REST API base URL.
    fn api_url(&self, path: &str) -> String {
        format!(
            "{}/api/v4{}",
            self.config.server_url.trim_end_matches('/'),
            path
        )
    }

    /// WebSocket URL.
    fn ws_url(&self) -> String {
        let base = self.config.server_url.trim_end_matches('/');
        let scheme = if base.starts_with("https") {
            "wss"
        } else {
            "ws"
        };
        let host = base
            .trim_start_matches("https://")
            .trim_start_matches("http://");
        format!("{}://{}/api/v4/websocket", scheme, host)
    }

    /// Convert a Mattermost post to a UnifiedMessage.
    fn post_to_unified(post: &MmPost, username: &str) -> UnifiedMessage {
        let text = &post.message;
        let is_cmd = text.starts_with('/');

        let content = if is_cmd {
            let parts: Vec<&str> = text[1..].split_whitespace().collect();
            let cmd = parts.first().unwrap_or(&"").to_string();
            let args: Vec<String> = parts[1..].iter().map(|s| s.to_string()).collect();
            MessageContent::command(text, cmd, args)
        } else {
            MessageContent::text(text)
        };

        let sender = Identity::new(&post.user_id).with_username(username.to_string());

        let thread_id = if post.root_id.is_empty() {
            None
        } else {
            Some(post.root_id.clone())
        };

        UnifiedMessage {
            id: post.id.clone(),
            channel: "mattermost".to_string(),
            sender,
            content,
            timestamp: chrono::DateTime::from_timestamp_millis(post.create_at)
                .unwrap_or_else(Utc::now),
            thread_id,
            reply_to_id: None,
            chat_id: Some(post.channel_id.clone()),
            raw: None,
            metadata: Default::default(),
        }
    }
}

#[async_trait]
impl ChannelAdapter for MattermostAdapter {
    fn channel_id(&self) -> &str {
        "mattermost"
    }

    async fn connect(&mut self) -> Result<()> {
        // Verify auth by fetching current user
        let resp = self
            .http
            .get(&self.api_url("/users/me"))
            .header(
                "Authorization",
                format!("Bearer {}", self.config.access_token),
            )
            .send()
            .await
            .map_err(|e| Error::Connection(format!("mattermost auth: {}", e)))?;

        if !resp.status().is_success() {
            return Err(Error::Connection(format!(
                "mattermost auth failed: HTTP {}",
                resp.status()
            )));
        }

        let user: MmUser = resp
            .json()
            .await
            .map_err(|e| Error::Connection(format!("mattermost user parse: {}", e)))?;
        self.user_id = Some(user.id.clone());

        // TODO: Establish WebSocket connection for real-time events
        // 1. Connect to ws_url()
        // 2. Send auth challenge with access_token
        // 3. Listen for "posted" events
        // 4. Parse post JSON and convert to UnifiedMessage
        // 5. Call handler

        let (tx, _rx) = tokio::sync::oneshot::channel();
        self.ws_shutdown = Some(tx);
        self.connected = true;
        Ok(())
    }

    async fn disconnect(&mut self) -> Result<()> {
        if let Some(tx) = self.ws_shutdown.take() {
            let _ = tx.send(());
        }
        self.connected = false;
        Ok(())
    }

    fn on_message(&mut self, handler: MessageHandler) {
        let h = Arc::clone(&self.handler);
        tokio::task::block_in_place(|| {
            let rt = tokio::runtime::Handle::current();
            rt.block_on(async {
                let mut guard = h.lock().await;
                *guard = Some(handler);
            });
        });
    }

    async fn send(&self, msg: OutboundMessage) -> Result<Option<String>> {
        let post = MmCreatePost {
            channel_id: msg.chat_id.clone(),
            message: msg.text.clone(),
            root_id: msg.thread_id.clone().or(msg.reply_to_id.clone()),
        };

        let resp = self
            .http
            .post(&self.api_url("/posts"))
            .header(
                "Authorization",
                format!("Bearer {}", self.config.access_token),
            )
            .json(&post)
            .send()
            .await
            .map_err(|e| Error::Send(format!("mattermost send: {}", e)))?;

        if !resp.status().is_success() {
            let body = resp.text().await.unwrap_or_default();
            return Err(Error::Send(format!("mattermost send failed: {}", body)));
        }

        let response: MmCreatePostResponse = resp
            .json()
            .await
            .map_err(|e| Error::Send(format!("mattermost response parse: {}", e)))?;

        Ok(Some(response.id))
    }

    async fn get_status(&self) -> ChannelStatus {
        ChannelStatus {
            connected: self.connected,
            channel: "mattermost".to_string(),
            account_id: self.user_id.clone(),
            error: None,
            last_activity: self.last_activity,
        }
    }
}
